src/lib/SaferNames/Name.hs

```diff
@@ -0,0 +1,478 @@
+type B = S -> S -> *  -- binder-y things, covariant in the first param and
+                      -- contravariant in the second
+
+-- A `FreshExt n l` means that `l` extends `n` with only fresh names
+data FreshExt (n::S) (l::S) = UnsafeMakeExt
+
+emptyFreshExt :: FreshExt n n
+emptyFreshExt = UnsafeMakeExt
+
+composeFreshExt :: FreshExt n h -> FreshExt h l -> FreshExt n l
+composeFreshExt _ _ = UnsafeMakeExt
+
+data NameTraversal (e::E) (m:: * -> *) (i::S) (o::S) where
+  NameTraversal :: (Name i -> m (e o))
+                -> RenameEnv (i:=>:i') o
+                -> NameTraversal e m i' o
+
+type RenameTraversal = NameTraversal Name
+
+-- traverses free names, possibly renaming bound ones as needed
+class HasNamesE (e::E) where
+  traverseNamesE :: Monad m => Scope o -> RenameTraversal m i o -> e i -> m (e o)
+
+-- A fresh binder, hiding its new output scope and exposing a renaming
+-- env that can give you those names. Used mainly for the result of
+-- `traverseNamesB` and similar.
+data FreshBinder (b::B) (n::S) (i::S) where
+  FreshBinder :: FreshExt o o' -> b o o' -> RenameEnv i o'
+              -> FreshBinder b o i
+
+class HasNamesB (b::B) where
+  traverseNamesB :: Monad m => Scope o -> RenameTraversal m i o
+                       -> b i i'
+                       -> m (FreshBinder b o (i:=>:i'))
+  boundScope :: b n l -> Scope (n:=>:l)
+
+-- slightly safer than raw `unsafeCoerce` because at least it checks the kind
+unsafeCoerceE :: forall (e::E) i o . e i -> e o
+unsafeCoerceE = unsafeCoerce
+
+unsafeCoerceB :: forall (b::B) n l n' l' . b n l -> b n' l'
+unsafeCoerceB = unsafeCoerce
+
+withFresh :: Scope n -> (forall l. FreshExt n l -> PlainBinder n l -> Name l -> a) -> a
+withFresh (UnsafeMakeEnv scope) cont =
+  cont UnsafeMakeExt (UnsafeMakeBinder freshName) (UnsafeMakeName freshName)
+  where freshName = freshRawName "v" (LM.keysSet scope)
+
+freshRawName :: D.Tag -> S.Set RawName -> RawName
+freshRawName tag usedNames = D.Name D.GenName tag nextNum
+  where
+    nextNum = case S.lookupLT (D.Name D.GenName tag bigInt) usedNames of
+                Just (D.Name D.GenName tag' i)
+                  | tag' /= tag -> 0
+                  | i < bigInt  -> i + 1
+                  | otherwise   -> error "Ran out of numbers!"
+                _ -> 0
+    bigInt = (10::Int) ^ (9::Int)  -- TODO: consider a real sentinel value
+
+unitName :: Name UnitScope
+unitName = UnsafeMakeName "TheName"
+
+unitScope :: Scope UnitScope
+unitScope = UnsafeMakeEnv $ LM.singleton unitRawName ()
+  where UnsafeMakeName unitRawName = unitName
+
+envMapWithKey :: (Name i -> a -> b) -> Env i a -> Env i b
+envMapWithKey f (UnsafeMakeEnv m) = UnsafeMakeEnv $ LM.mapWithKey f' m
+  where f' rawName = f $ UnsafeMakeName rawName
+
+envPairs :: Env (n:=>:l) a -> (PlainBinderList n l, [a])
+envPairs (UnsafeMakeEnv m) = do
+  let (names, vals) = unzip $ LM.assocs m
+  (unsafeMakePlainBinderList names, vals)
```

Comment:
Yes, an alternative would be something like,

```
evnPairs :: Env (n:=>:l) a -> Nest (AnnBinderP PlainBinder (LiftE a)) n l`
```
But unpacking nests is a pain because you have to worry about substituting the annotation at each step and so on. In this case, since the `a`s are actually all the same type (rather than each one having a slightly different scope) it seems more pain than it's worth. The lack of precision in the lengths seems like a small price to pay.